### ISSUE TITLE
pam_systemd_home requires root permissions

### ISSUE THREAD

**Issue by yellowsquid** [systemd/systemd#15072]
**systemd version the issue has been seen with**
245

**Used distribution**
Arch

**Expected behaviour you didn't see**
PAM authentication should succeed without root permissions.

**Unexpected behaviour you saw**
User authentication fails with the following message in the system journal:

```
pam_systemd_home(test-service:auth): Failed to acquire home for user test-user: Access denied
```

The problem doesn't prevent use of `sudo` or `su` because they have `setuid` bits.

**Steps to reproduce the problem**

Outline:
- Create a test PAM service.
- Create a test user using `homectl` with a LUKS encrypted home directory.
- Log in as the test user.
- Attempt to authenticate with the test service.

NOTE: you must run this all as root or you'll need functioning systemd-home authentication first. A working example can be found at the end of the man page for `pam_systemd_home`.

You will also need [`pamtester`] for this demonstration. It should be available for most distributions.

[`pamtester`]: http://pamtester.sourceforge.net/

Create `/etc/pam.d/test-service` with the following contents.

```
auth required pam_systemd_home.so
```

Create a test user, switch to them, and try to authenticate.
```
homectl create test-user --storage=luks --disk-size=100M
sudo -u test-user pamtester test-service test-user authenticate
```

This last command will fail with the following message on the console:
```
pamtester: Error in service module
```

Clean up.
```
rm /etc/pam.d/test-service
homectl remove test-user
```

**Comment by doverride** [comment]
Yes I noticed this well, this  breaks `vlock`

**Comment by soredake** [comment]
Screenlocker in kde is broken, says "Unlocking failed" and in sddm log "Authentication failure"

**Comment by Raimu-san** [comment]
`swaylock` is broken as well, I'm also on arch with systemd 245

**Comment by poettering** [comment]
hmm, so PAM auth is supposed to work without privs? is that documented anywhere, or is that just what people assume?

**Comment by yellowsquid** [comment]
I think it's just an assumption. pam_unix is going to be an almost globally used module, and it also makes this assumption. I think working without privileges will help with migration.

We also have to consider that if this module requires privileges, then we have to make every end application acquire them. In this thread alone, there are already three apps that need privileges. The same will be true for all screen lockers. The [Arch wiki] lists 21 currently. From a security point of view, making this module work without privileges will significantly reduce the attack vectors.

[Arch wiki]: https://wiki.archlinux.org/index.php/List_of_applications#Screen_lockers

**Comment by poettering** [comment]
well, reading /etc/shadow from unpriv code to perform pam_unix auth kinda sucks. setuid binaries are a bad think, not a good thing, security-wise.

**Comment by yellowsquid** [comment]
I agree that setuid is a bad idea. If there is a way to get around it then we should definitely try to do that. If there is not, then it is a better plan to only make a part of this module setuid than all of the apps that use pam for authentication. 

First we'd have to identify what exactly needs privileges and then extract only that into a separate binary. This is what pam_unix does (at least on Arch) with the unix_chkpwd binary.

A pam module that absolutely requires privileges isn't a new idea. I was suggesting that we follow prior work. Apologies if my argument didn't convey this.

**Comment by basedtho** [comment]
how does gdm do it?

**Comment by Raimu-san** [comment]
@basedtho I'm not entirely sure but I think that on GNOME, unlike the other environments, the screenlocker is actually GDM and therefore it has root access.

**Comment by poettering** [comment]
Fix waiting in #15737

**Comment by basedtho** [comment]
please backport this to 245
